### PR TITLE
Improve attention masking

### PR DESCRIPTION
**Pull request by UFO-101**
# Description

**Allow `attention_mask` to override the default mask in `HookedTransformer.forward()`.**
Currently the `attention_mask` argument is a footgun because it does nothing if `start_at_layer` is not `None`. One solution could be to just assert that `start_at_layer` is `None` when `attention_mask` is not `None`. But it is useful to be able to control the attention masking directly, so instead this just makes it do the thing you would expect.

**Add `attention_mask` argument to `loss_fn()` and `lm_cross_entropy_loss()` and adjust the cross entropy calculation to ignore masked (padding) tokens.**
Currently the loss calculation will be wrong if there are any padding tokens (or other masked tokens). This fixes that.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

**Comment by bryce13950** [comment]
Hey @UFO-101! Thanks for putting this together. I need to sit down and play around with this given that it is a change at one of the most important parts of the code base. At a glance it looks good, but I want to play around with the branch before I merge this. If anything comes up I will let you know.
